### PR TITLE
Bucket Replication APIs(Set, Get, Remove)

### PR DESCRIPTION
**Pull request by BigUstad**
Args, Examples

### CODE REVIEW

**Review by blaquezzz — APPROVED** [review]

**Review by sinhaashish — APPROVED** [review]

File: Minio/ApiEndpoints/IBucketOperations.cs [hunk 1]
```diff
@@ -1,6 +1,6 @@
 ﻿/*
  * MinIO .NET Library for Amazon S3 Compatible Cloud Storage,
- * (C) 2017, 2018, 2019 MinIO, Inc.
+ * (C) 2017, 2018, 2019, 2020, 2021 MinIO, Inc.
```
Comment (poornas) [inline Minio/ApiEndpoints/IBucketOperations.cs:3]:
```suggestion
 * (C) 2017-2021 MinIO, Inc.
```

File: Minio/DataModel/BucketOperationsReponse.cs [hunk 2]
```diff
@@ -327,4 +327,23 @@ internal GetObjectLockConfigurationResponse(HttpStatusCode statusCode, string re
             }
         }
     }
+
+    internal class GetBucketReplicationResponse : GenericResponse
+    {
+        internal ReplicationConfiguration BucketReplication { set; get; }
```
Comment (poornas) [inline Minio/DataModel/BucketOperationsReponse.cs:333]:
```suggestion
        internal ReplicationConfiguration Config { set; get; }
```

File: Minio/DataModel/BucketOperationsReponse.cs [hunk 3]
```diff
@@ -327,4 +327,23 @@ internal GetObjectLockConfigurationResponse(HttpStatusCode statusCode, string re
...
+        {
+            if (string.IsNullOrEmpty(responseContent) ||
+                    !HttpStatusCode.OK.Equals(statusCode))
+            {
+                this.BucketReplication = null;
+                return;
+            }
+            using (var stream = new MemoryStream(System.Text.Encoding.UTF8.GetBytes(responseContent)))
+            {
+                this.BucketReplication = (ReplicationConfiguration)new XmlSerializer(typeof(ReplicationConfiguration)).Deserialize(stream);
```
Comment (poornas) [inline Minio/DataModel/BucketOperationsReponse.cs:345]:
```suggestion
                this.Config = (ReplicationConfiguration)new XmlSerializer(typeof(ReplicationConfiguration)).Deserialize(stream);
```

File: Minio/ApiEndpoints/BucketOperations.cs [hunk 4]
```diff
@@ -497,6 +497,50 @@ public async Task RemoveObjectLockConfigurationAsync(RemoveObjectLockConfigurati
...
+        /// <param name="args">GetBucketReplicationArgs Arguments Object with information like Bucket name</param>
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
+        /// <returns>Replication configuration object</returns>
+        public async Task<ReplicationConfiguration> GetBucketReplicationAsync(GetBucketReplicationArgs args, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            args.Validate();
+            RestRequest request = await this.CreateRequest(args).ConfigureAwait(false);
+            IRestResponse restResponse = await this.ExecuteAsync(this.NoErrorHandlers, request, cancellationToken).ConfigureAwait(false);
+            GetBucketReplicationResponse response = new GetBucketReplicationResponse(restResponse.StatusCode, restResponse.Content);
+            return response.BucketReplication;
```
Comment (poornas) [inline Minio/ApiEndpoints/BucketOperations.cs:512]:
```suggestion
            return response.Config;
```

File: Minio/DataModel/Replication/DeleteReplication.cs [hunk 5]
```diff
@@ -0,0 +1,45 @@
...
+{
+    [Serializable]
+    [XmlRoot(ElementName = "DeleteReplication")]
+
+    public class DeleteReplication
+    {
+        [XmlElement("Status")]
+        public string Status { get; set; }
+        public const string DELETEREPLICATION_STATUS_ENABLED = "Enabled";
+        public const string DELETEREPLICATION_STATUS_DISABLED = "Disabled";
```
Comment (poornas) [inline Minio/DataModel/Replication/DeleteReplication.cs:30]:
```suggestion
        public const string StatusEnabled = "Enabled";
        public const string StatusDisabled = "Disabled";
```

File: Minio/DataModel/Replication/DeleteMarkerReplication.cs [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+{
+    [Serializable]
+    [XmlRoot(ElementName = "DeleteMarkerReplication")]
+
+    public class DeleteMarkerReplication
+    {
+        [XmlElement("Status")]
+        public string Status { get; set; }
+        public const string REPLICATION_DELETEMARKER_STATUS_ENABLED = "Enabled";
+        public const string REPLICATION_DELETEMARKER_STATUS_DISABLED = "Disabled";
```
Comment (poornas) [inline Minio/DataModel/Replication/DeleteMarkerReplication.cs:30]:
```suggestion
        public const string StatusEnabled = "Enabled";
        public const string StatusDisabled = "Disabled";
```

File: Minio/DataModel/Replication/ReplicationRule.cs [hunk 7]
```diff
@@ -0,0 +1,87 @@
...
+        [XmlElement(ElementName = "Prefix", IsNullable = true)]
+        public string Prefix { get; set; }
+        [XmlElement("DeleteReplication")]
+        public DeleteReplication DeleteReplication { get; set; }
+        [XmlElement(ElementName = "SourceSelectionCriteria", IsNullable = true)]
+        public SourceSelectionCriteria SourceSelectionCriteria { get; set; }
+        [XmlElement("Status")]
+        public string Status { get; set; }
+        public const string REPLICATION_RULE_STATUS_ENABLED = "Enabled";
+        public const string REPLICATION_RULE_STATUS_DISABLED = "Disabled";
```
Comment (poornas) [inline Minio/DataModel/Replication/ReplicationRule.cs:49]:
```suggestion
        public const string StatusEnabled = "Enabled";
        public const string StatusDisabled = "Disabled";
```

**Review by poornas — CHANGES_REQUESTED** [review]
was this tested against minio and aws

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 8]
```diff
@@ -87,6 +88,10 @@ public class FunctionalTest
         private const string setObjectRetentionSignature = "Task SetObjectRetentionAsync(SetObjectRetentionArgs args, CancellationToken cancellationToken = default(CancellationToken))";
         private const string getObjectRetentionSignature = "Task<ObjectRetentionConfiguration> GetObjectRetentionAsync(GetObjectRetentionArgs args, CancellationToken cancellationToken = default(CancellationToken))";
         private const string clearObjectRetentionSignature = "Task ClearObjectRetentionAsync(ClearObjectRetentionArgs args, CancellationToken cancellationToken = default(CancellationToken))";
+        private const string getBucketReplicationSignature = "Task<Tagging> GetBucketTagsAsync(GetBucketTagsArgs args, CancellationToken cancellationToken = default(CancellationToken))";
+        private const string setBucketReplicationSignature = "Task SetBucketTagsAsync(SetBucketTagsArgs args, CancellationToken cancellationToken = default(CancellationToken))";
+        private const string deleteBucketReplicationSignature = "Task RemoveBucketTagsAsync(RemoveBucketTagsArgs args, CancellationToken cancellationToken = default(CancellationToken))";
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs:97]: Mistake happened in merging, I think. Will correct.

File: Docs/API.md [hunk 9]
```diff
@@ -1128,6 +1132,148 @@ catch(MinioException e)
...
+
+|Param   | Type	  | Description  |
+|:--- |:--- |:--- |
+| ``args``  | _GetBucketReplicationArgs_  | GetBucketReplicationArgs Argument Object with bucket name  |
+| ``cancellationToken``| _System.Threading.CancellationToken_ | Optional parameter. Defaults to default(CancellationToken) |
+
+
+| Return Type	  | Exceptions	  |
+|:--- |:--- |
+|  ``Task<ReplicationConfiguration>``: ReplicationConfiguration object which contains the Replication configuration details. | Listed Exceptions: |
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Docs/API.md]: Changed the formatting to align properly.

**Comment by BigUstad** [comment]
> was this tested against minio and aws

Yes, I tested it with the remote folder on play & s3.

File: Docs/API.md [hunk 10]
```diff
@@ -1128,6 +1132,148 @@ catch(MinioException e)
...
+__Example__
+
+
+```cs
+try
+{
+    // Get Bucket Replication for the bucket
+    var args = new GetBucketReplicationArgs()
+                        .WithBucket(bucketName);
+    var lfc = await minio.GetBucketReplicationAsync(args);
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Docs/API.md]: Changed variable name

File: Docs/API.md [hunk 11]
```diff
@@ -1128,6 +1132,148 @@ catch(MinioException e)
...
+__Example__
+
+
+```cs
+try
+{
+    // Set Replication configuration for the bucket
+    SetBucketReplicationArgs args = new SetBucketReplicationArgs()
+                                                .WithBucket(bucketName)
+                                                .WithConfiguration(lfc);
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Docs/API.md]: Changed variable name.

File: Minio.Examples/Cases/GetBucketReplication.cs [hunk 12]
```diff
@@ -0,0 +1,54 @@
...
+        public async static Task Run(MinioClient minio,
+                                    string bucketName = "my-bucket-name")
+        {
+            try
+            {
+                Console.WriteLine("Running example for API: GetBucketReplicationConfiguration");
+                var repl = await minio.GetBucketReplicationAsync(
+                    new GetBucketReplicationArgs()
+                        .WithBucket(bucketName)
+                );
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Examples/Cases/GetBucketReplication.cs:33]: Moved to align with constructor.

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 13]
```diff
@@ -1,6 +1,6 @@
 ﻿/*
 * MinIO .NET Library for Amazon S3 Compatible Cloud Storage,
-* (C) 2017, 2018, 2019, 2020 MinIO, Inc.
+* (C) 2017, 2018, 2019, 2020, 2021 MinIO, Inc.
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs]: Changed it in Functional test.

[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs]: But removed the test as it isn't possible to run in current status.

File: Minio/DataModel/BucketOperationsReponse.cs (hunk 3 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsReponse.cs]: Changed member variable name accordingly.

File: Minio/ApiEndpoints/BucketOperations.cs (hunk 4 above)
Comment (BigUstad) [reply Minio/ApiEndpoints/BucketOperations.cs]: Changed member variable name accordingly.

File: Minio/DataModel/Replication/DeleteReplication.cs (hunk 5 above)
Comment (BigUstad) [reply Minio/DataModel/Replication/DeleteReplication.cs]: Changed member variable name.

File: Minio/DataModel/Replication/DeleteMarkerReplication.cs (hunk 6 above)
Comment (BigUstad) [reply Minio/DataModel/Replication/DeleteMarkerReplication.cs]: Changed member variable names.

File: Minio/DataModel/Replication/ReplicationRule.cs (hunk 7 above)
Comment (BigUstad) [reply Minio/DataModel/Replication/ReplicationRule.cs]: Changed member variable names.

File: Minio/DataModel/BucketOperationsReponse.cs (hunk 2 above)
Comment (BigUstad) [reply Minio/DataModel/BucketOperationsReponse.cs]: Changed var namr

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 14]
```diff
@@ -3792,106 +3791,79 @@ internal async static Task ObjectRetentionAsync_Test1(MinioClient minio)
...
+            Expiration exp = new Expiration(DateTime.Now.AddYears(1));
+            var compareDate = new DateTime(DateTime.Now.Year, DateTime.Now.Month, DateTime.Now.Day, 0 ,0 ,0);
+            var expInDays = (compareDate.AddYears(1) - compareDate).TotalDays;
+
+            LifecycleRule rule1 = new LifecycleRule(null, "txt", exp, null,
+                new RuleFilter(null, "txt/", null),
+                null, null, LifecycleRule.LIFECYCLE_RULE_STATUS_ENABLED
+                );
+            rules.Add(rule1);
+            LifecycleConfiguration lfc = new LifecycleConfiguration(rules);
```
Comment (poornas) [inline Minio.Functional.Tests/FunctionalTest.cs]: are these added to this PR in error?

File: Minio.Examples/Cases/GetBucketReplication.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        public async static Task Run(MinioClient minio,
+                                    string bucketName = "my-bucket-name")
+        {
+            try
+            {
+                Console.WriteLine("Running example for API: GetBucketReplicationConfiguration");
+                var repl = await minio.GetBucketReplicationAsync(
+                    new GetBucketReplicationArgs()
+                                    .WithBucket(bucketName)
+                );
```
Comment (poornas) [inline Minio.Examples/Cases/GetBucketReplication.cs:33]: these examples need to be added in Program.cs in this folder

File: Minio.Functional.Tests/FunctionalTest.cs (hunk 14 above)
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs]: All changes to functional tests removed.

File: Minio.Examples/Cases/GetBucketReplication.cs (hunk 15 above)
Comment (BigUstad) [reply Minio.Examples/Cases/GetBucketReplication.cs:33]: These are now added.

**Review by poornas — APPROVED** [review]
LGTM

**Pull request merged**